src/com/google/enterprise/adaptor/filenet/FileNetAdaptor.java

```diff
@@ -91,11 +120,201 @@ Connection getConnection() {
         context.getSensitiveValueDecoder().decodeValue(password));
   }
 
+  private static DocId newDocId(Checkpoint checkpoint) {
+    return new DocId("pseudo/" + checkpoint);
+  }
+
+  private static DocId newDocId(Id id) {
+    return new DocId("guid/" + id);
+  }
+
   @Override
-  public void getDocIds(DocIdPusher pusher) {
+  public void getDocIds(DocIdPusher pusher) throws IOException,
+      InterruptedException {
+    pusher.pushRecords(Arrays.asList(
+        new Record.Builder(newDocId(new Checkpoint("type=document")))
+            .setCrawlImmediately(true).build()));
   }
 
   @Override
-  public void getDocContent(Request req, Response resp) {
+  public void getDocContent(Request req, Response resp) throws IOException,
+      InterruptedException {
+    DocId id = req.getDocId();
+    String[] idParts = id.getUniqueId().split("/", 2);
+    if (idParts.length != 2) {
+      logger.log(Level.FINE, "Invalid DocId: {0}", id);
+      resp.respondNotFound();
+      return;
+    }
+    switch (idParts[0]) {
+      case "pseudo":
+        Checkpoint checkpoint = new Checkpoint(idParts[1]);
+        switch (checkpoint.type) {
+          case "document":
+            documentTraverser.getDocIds(checkpoint, context.getDocIdPusher());
+            break;
+          default:
+            logger.log(Level.WARNING, "Unsupported type: " + checkpoint);
+            resp.respondNotFound();
+            break;
+        }
+        break;
+      case "guid":
+        documentTraverser.getDocContent(new Id(idParts[1]), req, resp);
+        break;
+      default:
+        logger.log(Level.FINE, "Invalid DocId: {0}", id);
+        resp.respondNotFound();
+        return;
+    }
+  }
+
+  @VisibleForTesting
+  static interface Traverser {
+    void getDocIds(Checkpoint checkpoint, DocIdPusher pusher)
+        throws IOException, InterruptedException;
+
+    void getDocContent(Id id, Request request, Response response)
+        throws IOException, InterruptedException;
+  }
+
+  private static class MockTraverser implements Traverser {
+    private final String idFormat = "{AAAAAAAA-0000-0000-0000-%012d}";
+    private final int maxFeedUrls;
+
+    MockTraverser(int maxFeedUrls) {
+      this.maxFeedUrls = maxFeedUrls;
+    }
+
+    @Override
+    public void getDocIds(Checkpoint checkpoint, DocIdPusher pusher)
+        throws IOException, InterruptedException {
+      checkNotNull(checkpoint, "checkpoint may not be null");
+      checkNotNull(pusher, "pusher may not be null");
+      Date timestamp;
+      int id;
+      if (checkpoint.guid == null) {
```

Comment:
I've been trying to remember to key off the timestamp, which the primary column of the traversal. The GUID is alluring, but it's the secondary sort order, for example.

Maybe have the checkpoint itself answer this question in some fashion, e.g., Checkpoint.isEmpty()?